Refactor agentpoller to allow further commands

petrkotas opened the pull request:
**What this PR does / why we need it**:

Introduce concept of multiple pollers for agent.
Pollers can be simply added as additional worker and configured separately.

Introduce async cache for storing the agent data, this cache can be polled without penalty multiple times from different goroutines.

The goal of the cache is to enable async storage from multiple pollers.
It also enabled async access for future guest agent subresource. 
The subresource will directly access cached data and will not require 
waiting for the poller.

/cc @phoracek 

Signed-off-by: Petr Kotas <petr.kotas@gmail.com>

**Special notes for your reviewer**:
This is the preparation PR for introducing the guest agent data as subresource.
Original PR: #2827 

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

petrkotas commented:
/retest

petrkotas commented:
/retest

petrkotas commented:
@phoracek I did the refactoring as you requested, would you mind looking at this PR? Thank you.

petrkotas commented:
@stu-gott would you please take a look at the refactoring PR? It would really help. Thank you.

vatsalparekh commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller_test.go:
```diff
@@ -29,15 +29,77 @@ import (
+			// agentPoller.domainData = &DomainData{
+			// 	// net2 only present in DomainData
+			// 	aliasByMac: map[string]string{"0a:58:0a:f4:00:51": "ovs", "02:00:00:b0:17:66": "net1", "02:11:11:b0:17:66": "net2"},
+			// }
```
forgot to remove comments from here?

petrkotas replied:
You are right, deleted!

phoracek commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -0,0 +1,153 @@
+)
+
+// GuestOsInfo is the response from 'guest-get-osinfo'
+type GuestOsInfo struct {
```
Please keep this PR limited to change from simple to parallel poller. Are these new structures essential for the PR?

petrkotas replied:
It is only moved from `agent_poller.go`. Not introduced.

petrkotas commented on pkg/virt-launcher/notify-client/client.go:
```diff
@@ -197,8 +197,9 @@ func eventCallback(c cli.Connection, domain *api.Domain, libvirtEvent libvirtEve
 				events <- event
 			}
 		}
+		log.Log.Infof("7) OSINFO IN EVENT CALLBACK: %v", osInfo)
```
It is because I find it easier to locate caps in the logs. Thanks, I missed this log.

petrkotas commented on pkg/virt-launcher/notify-client/client.go:
```diff
@@ -240,11 +242,16 @@ func (n *Notifier) StartDomainNotifier(domainConn cli.Connection, deleteNotifica
+				interfaceStatuses = agentUpdate.DomainInfo.Interfaces
+				guestOsInfo = agentUpdate.DomainInfo.OSInfo
+				if domainCache != nil && interfaceStatuses != nil {
+					// do merge here
```
Thanks, removed this one.

ksimon1 commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -0,0 +1,153 @@
+	guestOSInfo := GuestOsInfo{}
+	response := stripAgentResponse(agentReply)
+
+	json.Unmarshal([]byte(response), &guestOSInfo)
```
What about error handling here?

petrkotas replied:
Excellent eye, I have totally missed the error check.

ksimon1 commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -0,0 +1,153 @@
+	interfaces := []Interface{}
+	response := stripAgentResponse(agentReply)
+
+	json.Unmarshal([]byte(response), &interfaces)
```
same here

petrkotas replied:
The same.

petrkotas commented:
Hi @phoracek @ksimon1 and @vatsalparekh I have made the changes that have been requested. Would you please take another look?

ksimon1 commented:
LGTM

mhenriks commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -31,265 +30,183 @@ import (
-	Type   string `json:"ip-address-type"`
-	Prefix int    `json:"prefix"`
-}
+	s.store.Store(key, value)
```
Another thread could theoretically store the same key (and value) between calls to Load/Store here.  Which would lead to multiple events.  Probably not a big deal but a single mutex around both calls could protect from that.  Not currently an issue because only one poller thread.

dhiller replied:
Hmm, wouldn't that actually be what we want? I mean, if multiple changes arise one after another, we wouldn't want to skip a change, or would we? Asked another way, would we want to prohibit changes occuring nearly at the same time, and if so, why?

mhenriks replied:
I think we may want to prevent getting two consecutive events for the same key+value.  And that is currently possible.

petrkotas replied:
Generating two consecutive call require two poll workers with the same configuration polling the same guest agent endpoint. Multiple pollers with the same guest agent endpoint is not the goal. There will be multiple workers, indeed. However each one will have different set of guest agent endpoints, and each will have different timeout.

There will be 4 workers, taken from PR: #2827 

* sys data: INTERFACES, OSINFO, TIMEZONE, HOSTNAME; queried every: 120s
* Agent: AGENT_VERSION; queried every: 300s
* Filesystem: FILESYSTEM; queried every: 300s
* User list: USERS: queried every: 10s

Enabling user to configure the choice of distribution does not make sense.
The calls are expensive and timeouts are chosen by trial in oVirt to not stress guest agent.

Furthermore timeouts can be tuned in PR: #2827 via parameter to virt-launcher, but I would suggest to not touch it.

The rule of thumb is, timeouts are selected to reflect how often the data changes on the host.

mhenriks commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -31,265 +30,183 @@ import (
+	// this have to be done via configuration passed in, for now this is OK
+	p.workers = append(p.workers, PollerWorker{
+		CallTick:      qemuAgentPollerInterval,
+		AgentCommands: []AgentCommand{GET_INTERFACES, GET_OSINFO},
```
Would it make sense to have these as a parameter to `CreatePoller`?

I also assume that if there are ever multiple pollers, they will not be handling the same commands?

petrkotas replied:
Enabling the full configuration does not make sense. There will be only 4 polling workers. Each worker have its own special group (inherited from oVirt). Next is taken from PR: #2827 

* sys data: INTERFACES, OSINFO, TIMEZONE, HOSTNAME; queried every: 120s
* Agent: AGENT_VERSION; queried every: 300s
* Filesystem: FILESYSTEM; queried every: 300s
* User list: USERS: queried every: 10s

Enabling user to configure the choice of distribution does not make sense.
The calls are expensive and timeouts are chosen by trial in oVirt to not stress guest agent.

Also. These pollers are mutually exclusive and does not affect each other. This is the reason, why I selected the `async.Map`. Taken from the documentation:
```
The Map type is optimized for two common use cases: 
(1) when the entry for a given key is only ever written once 
but read many times, as in caches that only grow, or 
(2) when multiple goroutines read, write, and overwrite entries for
disjoint sets of keys. In these two cases, use of a Map may 
significantly reduce lock contention compared to 
a Go map paired with a separate Mutex or RWMutex. 
``` 
Case present in this PR and PR: #2827 is case (2).

dhiller commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller_test.go:
```diff
@@ -29,15 +29,71 @@ import (
+		It("should merge QEMU info and agent info", func() {
+			// agentPoller := AgentPoller{}
+
+			json_input := "{\"return\":[{\"name\":\"lo\",\"ip-addresses\":[{\"ip-address-type\":\"ipv4\",\"ip-address\":\"127.0.0.1\",\"prefix\":8},{\"ip-address-type\":\"ipv6\",\"ip-address\":\"::1\",\"prefix\":128}],\"hardware-address\":\"00:00:00:00:00:00\"},{\"name\":\"eth0\",\"ip-addresses\":[{\"ip-address-type\":\"ipv4\",\"ip-address\":\"10.244.0.81\",\"prefix\":24},{\"ip-address-type\":\"ipv6\",\"i ...
```
This is really hard to read. Being able to read this as plain multiline json without lots of escaped double quotes would really help me to have a more clear view of what's going on here. Maybe you could convert the test data into a multi line string?

petrkotas replied:
You are right, fixed that in the whole test file. Thanks for pointing it up, I was so used to it, that it did not occurred to me.

dhiller commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller_test.go:
```diff
@@ -87,5 +142,51 @@ var _ = Describe("Qemu agent poller", func() {
+				}
+
+				agentStore.Store(GET_OSINFO, fakeInfo)
+				Expect(agentStore.AgentUpdated).To(Receive(Equal(AgentUpdatedEvent{
```
This check is redundant (see above). Do you do this because the channel has to be polled? Or can we just skip the event somehow?

petrkotas replied:
This check have to be present. The second poll is required to see whether the event has not been fired, and this is done via additional poll.

dhiller replied:
I understand that the polling is required. I'm just complaining about the duplicate check. Well, that's just nitpicking so nevermind.

dhiller commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+// calculateInterfaceStatusesFromAgentJson does the conversion from agent info to api domain interfaces
+func calculateInterfaceStatusesFromAgentJson(agentResult []Interface) []api.InterfaceStatus {
```
nit: this is just a conversion, so the method should be named `convert...`

dhiller commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -0,0 +1,162 @@
+	return resultInterfaces
+}
+
+// MergeAgentStatusesWithDomainData merge QEMU interfaces with agent interfaces
```
nit: s/merge/merges/g

dhiller commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -31,265 +30,183 @@ import (
-	Machine       string `json:"machine"`
-	Id            string `json:"id"`
+// NewAgentAstore creates new agent store
+func NewAgentAstore() AgentStore {
```
Why is it `AgentAstore`? My guess: you mean `AsyncAgentStore`? Then either you should write `NewAgentAStore`, but `NewAsyncAgentStore` would be more clear.

dhiller commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -0,0 +1,162 @@
+
+	err := json.Unmarshal([]byte(response), &guestOSInfo)
+	if err != nil {
+		log.Log.Errorf("Cannot parse guestosinfo reply: %s", err.Error())
```
If the error was returned, this would enable better unit testing by being able to check for error conditions.

dhiller commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -0,0 +1,162 @@
+
+	err := json.Unmarshal([]byte(response), &interfaces)
+	if err != nil {
+		log.Log.Errorf("Cannot parse interface reply: %s", err.Error())
```
If the error was returned, this would enable better unit testing by being able to check for error conditions.

dhiller commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -0,0 +1,160 @@
+	return interfaceStatuses
+}
+
+// calculateInterfaceStatusesFromAgentJson does the conversion from agent info to api domain interfaces
```
Please adjust method name here also.

dhiller commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller_test.go:
```diff
@@ -28,16 +28,161 @@ import (
 
 var _ = Describe("Qemu agent poller", func() {
 	Context("recieving a reply from the agent", func() {
+		JSONInput := `{
```
Much better! :+1:

petrkotas commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -0,0 +1,162 @@
+// embeded object.
+// It is a workaround so the amount of copy paste code is limited
+func stripAgentResponse(agentReply string) string {
+	re := regexp.MustCompile(`{\s*\"return\":\s*([{\[][\s\S]*[}\]])\s*}`)
```
Thanks, fixed that.

petrkotas commented:
Thank you @mhenriks and @dhiller for thorough review. It helped a lot to polish the details. I believe I have addressed the comments. Once you have a time, would be able to do a quick look again? Thank you!

petrkotas commented:
Regarding my IDE, it is VSCode and it does not have spell check by default. However, thanks to you @dhiller I searched for an extension and I have a spellcheck.

dhiller commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -31,265 +30,182 @@ import (
-// Result for json unmarshalling
-type Result struct {
-	Interfaces []Interface `json:"return"`
+// AgentStore stores the agent data converted to api domain objects
```
Please adjust comment also

dhiller commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -31,265 +30,182 @@ import (
-	KernelVersion string `json:"kernel-version"`
-	Machine       string `json:"machine"`
-	Id            string `json:"id"`
+// NewAgentAstore creates new agent store
```
Please adjust comment.

dhiller commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -0,0 +1,160 @@
+}
+
+// calculateInterfaceStatusesFromAgentJson does the conversion from agent info to api domain interfaces
+func convertInterfaceStatusesFromAgentJson(agentResult []Interface) []api.InterfaceStatus {
```
Another nit: s/Json/JSON/

dhiller commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -0,0 +1,160 @@
+	return interfaceStatuses
+}
+
+func extractIps(ipAddresses []IP) (string, []string) {
```
nit: s/Ips/IPs/

dhiller commented on pkg/virt-launcher/notify-client/client.go:
```diff
@@ -177,7 +177,7 @@ func eventCallback(c cli.Connection, domain *api.Domain, libvirtEvent libvirtEve
 		}
 
-		log.Log.Infof("kubevirt domain status: %v(%v):%v(%v)", domain.Status.Status, status, domain.Status.Reason, reason)
+		log.Log.V(5).Infof("kubevirt domain status: %v(%v):%v(%v)", domain.Status.Status, status, domain.Status.Reason, reason)
```
I understand that default for verbosity is 2. Why do you reduce the log verbosity here? I think that in general it would help debugging this in case anything goes wrong here?

dhiller commented on pkg/virt-launcher/notify-client/client.go:
```diff
@@ -254,7 +256,6 @@ func (n *Notifier) StartDomainNotifier(domainConn cli.Connection, deleteNotifica
 	}()
 
 	domainEventLifecycleCallback := func(c *libvirt.Connect, d *libvirt.Domain, event *libvirt.DomainEventLifecycle) {
-		log.Log.Infof("DomainLifecycle event %d with reason %d received", event.Event, event.Detail)
```
Why is the logging removed here?

dhiller commented on pkg/virt-launcher/notify-client/client.go:
```diff
@@ -272,7 +273,7 @@ func (n *Notifier) StartDomainNotifier(domainConn cli.Connection, deleteNotifica
 	}
 
 	agentEventLifecycleCallback := func(c *libvirt.Connect, d *libvirt.Domain, event *libvirt.DomainEventAgentLifecycle) {
-		log.Log.Infof("GuestAgentLifecycle event state %d with reason %d received", event.State, event.Reason)
```
Why is the verbosity reduced here?

petrkotas commented:
Ouu. Thanks @dhiller the logs you discovered were actually casualties of my debug log cleanup. I switched to robot phase and just deleted or raised verbosity for every log I touched.

Thanks!

dhiller commented:
/lgtm

danielBelenky commented:
/lgtm

danielBelenky commented:
/approve
